### DISCUSSION TITLE
RPI RCOS Projects

### DISCUSSION THREAD

**Discussion by johnwason** [robotraconteur/robotraconteur#271] [General]
Students participating in the RPI Rensselaer Center for Open Source (RCOS) program (and other student open source developers!) are invited to participate in the Robot Raconteur project ecosystem. The Robot Raconteur ecosystem consists of the core implementations of the Robot Raconteur framework, and a number of drivers and utilities. The [directory](https://github.com/robotraconteur/robotraconteur-directory) contains a list of the various projects in the ecosystem.

The Robot Raconteur ecosystem is under active development, and there are several project opportunities for students to contribute:

* Robot Raconteur Diagnostic Utility
* Lua Bindings for Robot Raconteur Core
* PyRI Open Source Teach Pendant
* General Ecosystem Development

See the comments below for more details on each project and to claim your spot!

The RPI ECSE department has agreed to make a [ReThink Baxter](https://en.wikipedia.org/wiki/Baxter_(robot)) robot available to RPI students participating in Robot Raconteur development!

**Star this repository!** Every star helps get the word out!

**Comment by johnwason** [comment]
## Robot Raconteur Diagnostic Utility
A diagnostic utility is needed to help users configure, debug, and deploy projects consisting of multiple Robot Raconteur nodes. The utility will provide both a graphical user interface, and a command line interface so that it can be used on headless systems. It is suggested that the utility be developed in Python to provide maximum engagement with the Robot Raconteur community, which has limited software engineering skills beyond Python. The proposed utility would include the following features:
* Discovery browser to view available nodes and services available on network (Currently available through standalone “Service Browser” tool: https://github.com/robotraconteur/robotraconteur_servicebrowser)
* Service object browser to interact with the discovered services
  * View and pull the service definitions used by the service
  * View the attributes of the service
  * Browse the objects available within the service
  * Interact with object members in an interactive manner
  * Provide plugins for specific object types. For instance, custom panels for “Robot” and “Camera” object types
* Data recording and playback
  * Data plotting
  * 3D scene viewer

**Comment by johnwason** [comment]
## Lua Bindings for Robot Raconteur Core

Robot Raconteur Core is the primary implementation for the Robot Raconteur project and is developed in C++. Language “bindings” allow the core library to be used from other programming languages such as Python, Matlab, LabView, C#, C++, and Java. Lua is a lightweight scripting language that is used when a more resource intensive solution like Python is not required or possible. Lua is typically found in projects developed in C or C++ that need a lightweight scripting language.
* Develop Lua bindings by wrapping the Robot Raconteur Core library
* Use the Matlab bindings as a reference
* Primary challenge for the development will be converting between Robot Raconteur messages and Lua data types

**Comment by johnwason** [comment]
## PyRI Open Source Teach Pendant

The Python Restricted Industrial (PyRI) Open Source Teach Pendant (https://github.com/pyri-project/pyri-core) is a user interface developed as part of an Advanced Robotics for Manufacturing (ARM) Institute funded project. The project is designed to allow for novice users to access advanced technologies available in Robot Raconteur and ROS. The user interface provides scripting using Blockly and a restricted dialect of Python.
This project is still under development, with opportunities for student contribution:
* General project architecture and stability
* Database and data management design
* Front-end user interface design
* Programming language and runtime development for the restricted Python dialect. This task is more theoretical, involving the design of the restricted Python dialect itself
* Debugging integration with scripting sandbox and user interface
* Robot planning, trajectory generation, and 3D viewer

**Reply by masshu0504** [reply]
Hello! Im really interested in this project, I would love to learn more and how I could contribute. Please let me know if theres room. Thanks!

**Reply by johnwason** [reply]
Hi @masshu0504 so far you are the only student for Spring 2026

**Reply by johnwason** [reply]
Hi @masshu0504  send me a message at https://www.wasontech.com/contact-us

**Comment by johnwason** [comment]
## General Ecosystem Development

Have another idea? Let us known! Any contributions adding support for additional devices and software packages are welcome!

**Comment by OrangeTangy** [comment]
I am currently not working on the project, but you can talk to mr. John
Wason or email him regarding it.
